### PR TITLE
add sparseFullOuterJoin #545

### PR DESCRIPTION
**Pull request by nevillelyh**

### CODE REVIEW

File: scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala [hunk 1]
```diff
@@ -353,6 +352,39 @@ class PairSCollectionFunctions[K, V](val self: SCollection[(K, V)])
...
+   * Full outer join for cases when `this` is much larger than `that` which cannot fit in memory,
+   * but contains a mostly overlapping set of keys as `this`. A Bloom Filter of keys in `that` is
+   * used to split `this` into 2 partitions. Only those with keys in the filter go through the
+   * join and the rest are concatenated. This is useful for joining historical aggregates with
+   * incremental updates.
+   * @group join
+   * @param thatNumKeys estimated number of keys in `that`
+   * @param fpProb false positive probability when computing the overlap
+   */
+  def bloomFilterOuterJoin[W: ClassTag](that: SCollection[(K, W)],
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: Not sure about the name though. Some other candidates are: `unbalanced`, `lookup`, `incremental`. What else?

Comment (ravwojdyla) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: How about `sparseOuterJoin`, would be more consistent with `skewJoin`. `skew` is partitioning to get the hot keys, and `sparse` is partitioning to get the "cold" keys.

File: scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala [hunk 2]
```diff
@@ -353,6 +352,39 @@ class PairSCollectionFunctions[K, V](val self: SCollection[(K, V)])
...
+                                        thatNumKeys: Int,
```
Comment (ravwojdyla) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: `Int` => `Long`

File: scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala [hunk 3]
```diff
@@ -353,6 +352,39 @@ class PairSCollectionFunctions[K, V](val self: SCollection[(K, V)])
     hotJoined ++ chillJoined
   }
 
+  /**
```
Comment (ravwojdyla) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: Does it require to import any of `algebird` (like for `Hash128`), if so might be worth adding in the doc like in the `skewJoin`.

File: scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala [hunk 4]
```diff
@@ -353,6 +352,39 @@ class PairSCollectionFunctions[K, V](val self: SCollection[(K, V)])
...
+   * Full outer join for cases when `this` is much larger than `that` which cannot fit in memory,
```
Comment (ravwojdyla) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: Might be worth linking to the algebird doc for bloom filters for the `fpProb`.

File: scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala (hunk 2 above)
Comment (ravwojdyla) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: I guess not - cause `optimalWidth` wants Int anyway.

File: scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala (hunk 3 above)
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala]: No need to. `trait Hash128` has a companion `object Hash128` with common instances and those are in the default lookup scope. Nice little implicit trick.

**Pull request merged**